File: UnityScript/src/Scripts/OGWidget.js
#pragma strict

public enum RelativeX {
	Left,
	Center,
	Right
}
	
public enum RelativeY {
	Top,
	Center,
	Bottom
}

public enum ScreenSize {
	None,
	ScreenWidth,
	ScreenHeight
}

public enum ButtonAction {
	OnRelease,
	OnPress,
	OnHover
}

public class OGWidget extends MonoBehaviour {
	public class Stretch {
		public var width : ScreenSize = ScreenSize.None;
		public var widthFactor : float = 1.0;
		public var widthOffset : float = 0.0;
		
		public var height : ScreenSize = ScreenSize.None;
		public var heightFactor : float = 1.0;
		public var heightOffset : float = 0.0;
	}
	
	public class Pivot {
		public var x : RelativeX;
		public var y : RelativeY;
	}
	
	public class Anchor {		
		public var x : RelativeX = RelativeX.Left;
		public var xOffset : float = 0.0;
		
		public var y : RelativeY = RelativeY.Top;
		public var yOffset : float = 0.0;
	}

	public var isDrawn : boolean = true;
	public var isSelectable : boolean = false;
	public var isDraggable : boolean = false;
	public var resetAfterDrag : boolean = false;
	public var pivot : Pivot = new Pivot();
	public var anchor : Anchor = new Anchor();	
	public var stretch : Stretch = new Stretch();
	
	@HideInInspector public var styles : OGWidgetStyles = new OGWidgetStyles();
	@HideInInspector public var drawCrd : Rect;
	@HideInInspector public var drawRct : Rect;
	@HideInInspector public var clipRct : Rect;
	@HideInInspector public var mouseRct : Rect;
	@HideInInspector public var drawDepth : float;
	@HideInInspector public var scrollOffset : Vector3;
	@HideInInspector public var dragOffset : Vector3;
	@HideInInspector public var dragOrigPos : Vector3;
	@HideInInspector public var offset : Vector3;
	@HideInInspector public var hidden : boolean = false;
	@HideInInspector public var disabled : boolean = false;
	@HideInInspector public var outOfBounds : boolean = false;
	@HideInInspector public var isDirty : boolean = false;
	@HideInInspector public var inScrollView : boolean = false;
	@HideInInspector public var root : OGRoot;
	
	// TODO: Deprecate
	@HideInInspector public var styleName : String;
	
	
	//////////////////
	// Calculations
	//////////////////
	// Find child
	public function FindChild ( n : String ) : GameObject {
		for ( var i : int = 0; i < this.transform.childCount; i++ ) {
			if ( this.transform.GetChild ( i ).gameObject.name == n ) {
				return this.transform.GetChild ( i ).gameObject;
			}
		}
	
		return null;
	}

	// Combine rects
	public function CombineRects ( a : Rect, b : Rect ) : Rect {
		var result : Rect = new Rect ( 0, 0, 0, 0 );
		
		result.x = ( a.x < b.x ) ? a.x : b.x;
		result.y = ( a.y < b.y ) ? a.y : b.y;
		result.xMax = ( a.xMax > b.xMax ) ? a.xMax : b.xMax;
		result.yMax = ( a.yMax > b.yMax ) ? a.yMax : b.yMax;

		return result;
	}

	// Check mouseover
	public function CheckMouseOver () : boolean {
		if ( mouseRct != null ) {
			return CheckMouseOver ( mouseRct );
		} else {
			return false;
		}	
	}
	
	public function CheckMouseOver ( rect : Rect ) : boolean {
		var x : float = Input.mousePosition.x;
		var y : float = Input.mousePosition.y;
		
		return x > rect.x && y > rect.y && y < rect.y + rect.height && x < rect.x + rect.width;
	}
	
	public function CheckMouseOver ( rect1 : Rect, rect2 : Rect ) : boolean {
		return CheckMouseOver ( rect1 ) || CheckMouseOver ( rect2 );
	}
	
	// Coordinates (based on texture size)
	private function RecalcCoords ( coords : Rect ) : Rect {
		if ( this.GetType () == OGTexture ) {
			coords.x = 0;
			coords.y = 0;
			coords.width = 1;
			coords.height = 1;

		} else if ( root ) {
			coords.x /= root.texWidth;
			coords.y /= root.texHeight;
			coords.width /= root.texWidth;
			coords.height /= root.texHeight;
		}


		return coords;
	}
		
	// Scale (based on screen size)
	public function RecalcScale () : Vector3 {
		CalcStretch ();
		
		var scrollView : OGScrollView = this as OGScrollView;

		if ( scrollView ) {
			return new Vector3 ( scrollView.size.x, scrollView.size.y, 1 );
		} else {
			return this.transform.lossyScale;
		}
	}
	
	// Position (based on screen size)
	public function RecalcPosition () : Vector3 {
		CalcAnchor ();
		CalcPivot ();
		
		var newPos : Vector3 = this.transform.position;
		var scrollView : OGScrollView = this as OGScrollView;
		
		newPos += offset;
		newPos += scrollOffset;
		
		if ( scrollView ) {
			newPos.y += scrollView.size.y;	
		} else {
			newPos.y += this.transform.lossyScale.y;
		}

		newPos.y = Screen.height - newPos.y;
		
		return newPos;
	}
	
	// Stretch (based on screen size)
	private function CalcStretch () {
		if ( !stretch ) { return; }
		
		var newScale : Vector3 = this.transform.localScale;
		var scrollView : OGScrollView = this as OGScrollView;

		if ( scrollView ) {
			newScale.x = scrollView.size.x;
			newScale.y = scrollView.size.y;
		}
		
		if ( stretch.width == ScreenSize.ScreenWidth ) {
			newScale.x = ( Screen.width * stretch.widthFactor ) + stretch.widthOffset;
		} else if ( stretch.width == ScreenSize.ScreenHeight ) {
			newScale.x = ( Screen.height * stretch.widthFactor ) + stretch.widthOffset;
		}
		
		if ( stretch.height == ScreenSize.ScreenWidth ) {
			newScale.y = ( Screen.width * stretch.heightFactor ) + stretch.heightOffset;
		} else if ( stretch.height == ScreenSize.ScreenHeight ) {
			newScale.y = ( Screen.height * stretch.heightFactor ) + stretch.heightOffset;
		}
		
		if ( !scrollView ) {
			this.transform.localScale = newScale;
		
		} else {
			scrollView.size.x = newScale.x;
			scrollView.size.y = newScale.y;
		}
	
	}
	
	// Anchor (based on screen size)
	private function CalcAnchor () {
		if ( !anchor ) { return; }
		
		var newPos : Vector3 = this.transform.position;
		
		if ( anchor.x == RelativeX.Left ) {
			newPos.x = anchor.xOffset;
		} else if ( anchor.x == RelativeX.Center ) {
			newPos.x = ( Screen.width / 2 ) + anchor.xOffset;
		} else if ( anchor.x == RelativeX.Right ) {
			newPos.x = Screen.width + anchor.xOffset;
		}
		
		if ( anchor.y == RelativeY.Top ) {
			newPos.y = anchor.yOffset;
		} else if ( anchor.y == RelativeY.Center ) {
			newPos.y = ( Screen.height / 2 ) + anchor.yOffset;
		} else if ( anchor.y == RelativeY.Bottom ) {
			newPos.y = Screen.height + anchor.yOffset;
		}

		this.transform.position = newPos;
	}
	
	// Pivot (based on object size)
	private function CalcPivot () {
		switch ( pivot.y ) {
			case RelativeY.Top:
				offset.y = 0;
				break;
								
			case RelativeY.Center:
				offset.y = -this.transform.lossyScale.y/2;
				break;
				
			case RelativeY.Bottom:
				offset.y = -this.transform.lossyScale.y;
				break;
		}
		
		switch ( pivot.x ) {	
			case RelativeX.Right:
				offset.x = -this.transform.lossyScale.x;
				break;
						
			case RelativeX.Center:	
				offset.x = -this.transform.lossyScale.x/2;
				break;
			
			case RelativeX.Left:	
				offset.x = 0;
				break;
		}
	}
	
	// Calculate clipping
	public function CalcClipping ( tempRct : Rect ) {
		var shouldClip : boolean = clipRct.width > 0 && clipRct.height > 0;

		if ( shouldClip ) {
			var clipTop : float = (clipRct.y+clipRct.height) - (tempRct.y+tempRct.height);
			var clipRight : float = (clipRct.x+clipRct.width) - (tempRct.x+tempRct.width);
			var clipBottom : float = tempRct.y - clipRct.y;
			var clipLeft : float = tempRct.x - clipRct.x;
			
			if ( clipTop < 0 ) { tempRct.height += clipTop; }
			if ( clipRight < 0 ) { tempRct.width += clipRight; }
			if ( clipBottom < 0 ) { tempRct.y -= clipBottom; tempRct.height += clipBottom; }
			if ( clipLeft < 0 ) { tempRct.x -= clipLeft; tempRct.width +=clipLeft; }
		}

		return tempRct;
	}
	
	
	// Apply all calculations
	public function Recalculate () {
		if ( !styles.basic ) { return; }

		var drawScl : Vector3 = RecalcScale ();
		var drawPos : Vector3 = RecalcPosition ();
		drawCrd = RecalcCoords ( styles.basic.coordinates );
		drawDepth = -this.transform.position.z;
			
		var tempRct : Rect = new Rect ( drawPos.x, drawPos.y, drawScl.x, drawScl.y );
		if ( this.GetType() != OGLabel ) {
			tempRct = CalcClipping ( tempRct );	
		}
		drawRct = tempRct;
	}	

	
	//////////////////
	// Returns
	//////////////////
	public function GetRoot () : OGRoot {
		if ( !root ) {
			root = GameObject.FindObjectOfType.<OGRoot>();
		}
		
		return root;
	}
	
	
	//////////////////
	// Mouse
	//////////////////
	public function OnMouseDown () {};
	public function OnMouseUp () {};
	public function OnMouseOver () {};
	public function OnMouseDrag () {};
	public function OnMouseCancel () {};
	
	
	//////////////////
	// Update
	//////////////////
	public function Start () { SetDirty(); }
	public function OnEnable () { SetDirty(); }
	public function OnDisable () { SetDirty(); }
	public function OnDestroy () { SetDirty(); }
	public function Build () {}
	public function SetDrawn ( drawn : boolean ) {}
	public function UpdateWidget () {} 
	public function GetDefaultStyles () {
		GetRoot().skin.GetDefaultStyles ( this );
		Build ();
	}
	public function SetDirty () {
		if ( GetRoot() ) {
			GetRoot().SetDirty();
		}
	}

	//////////////////
	// Draw
	//////////////////
	public function DrawGL () {}
}

File: OpenGUI/Assets/Plugins/OpenGUI/Scripts/OGWidget.cs
using UnityEngine;
using System.Collections;

public enum RelativeX {
	None,
	Left,
	Center,
	Right,
	Factor
}
	
public enum RelativeY {
	None,
	Top,
	Center,
	Bottom,
	Factor
}

public enum ScreenSize {
	None,
	ScreenWidth,
	ScreenHeight
}

public enum ButtonAction {
	OnRelease,
	OnPress,
	OnHover
}

public class OGWidget : MonoBehaviour {
	[System.Serializable]
	public class Stretch {
		public ScreenSize width = ScreenSize.None;
		public float widthFactor = 1.0f;
		public float widthOffset = 0.0f;
		
		public ScreenSize height = ScreenSize.None;
		public float heightFactor = 1.0f;
		public float heightOffset = 0.0f;
	}
	
	[System.Serializable]
	public class Pivot {
		public RelativeX x;
		public RelativeY y;
	}
	
	[System.Serializable]
	public class Anchor {		
		public RelativeX x = RelativeX.None;
		public float xOffset = 0.0f;
		public float xFactor = 0.0f;
		
		public RelativeY y = RelativeY.None;
		public float yOffset = 0.0f;
		public float yFactor = 0.0f;
	}

	public OGRoot root;
	public bool isDrawn = true;
	public bool isDisabled = false;
	public bool isSelectable = false;
	public bool isDraggable = false;
	public bool resetAfterDrag = false;
	public Color tint = Color.white;
	public Pivot pivot = new Pivot();
	public Anchor anchor = new Anchor();	
	public Stretch stretch = new Stretch();
	public OGWidget clipTo;
	
	[HideInInspector] public OGStyleSet styles = new OGStyleSet();
	[HideInInspector] public OGStyle currentStyle;
	[HideInInspector] public Rect drawCrd;
	[HideInInspector] public Rect drawRct;
	[HideInInspector] public Rect mouseRct;
	[HideInInspector] public float drawDepth;
	[HideInInspector] public Vector3 scrollOffset;
	[HideInInspector] public Vector3 dragOffset;
	[HideInInspector] public Vector3 dragOrigPos;
	[HideInInspector] public Vector3 offset;
	[HideInInspector] public bool hidden = false;
	[HideInInspector] public bool outOfBounds = false;
	[HideInInspector] public bool isDirty = false;
	[HideInInspector] public bool isAlwaysOnTop = false;
	
	//////////////////
	// Calculations
	//////////////////
	// Convert to enum
	public OGWidgetType ToEnum () {
		return OGSkin.GetWidgetEnum ( this );
	}

	// Get scaled rect
	public Rect scaledRct {
		get {
			Rect result = drawRct;

			OGRoot root = OGRoot.GetInstance ();

			if ( root ) {
				result.x *= root.ratio.x;
				result.width *= root.ratio.x;
				result.y *= root.ratio.y;
				result.height *= root.ratio.y;
			}

			return result;
		}
	}

	// Find child
	public GameObject FindChild ( string n ) {
		for ( int i = 0; i < this.transform.childCount; i++ ) {
			if ( this.transform.GetChild ( i ).gameObject.name == n ) {
				return this.transform.GetChild ( i ).gameObject;
			}
		}
	
		return null;
	}

	// Combine rects
	public Rect CombineRects ( Rect a, Rect b ) {
		Rect result = new Rect ( 0, 0, 0, 0 );
		
		result.x = ( a.x < b.x ) ? a.x : b.x;
		result.y = ( a.y < b.y ) ? a.y : b.y;
		result.xMax = ( a.xMax > b.xMax ) ? a.xMax : b.xMax;
		result.yMax = ( a.yMax > b.yMax ) ? a.yMax : b.yMax;

		return result;
	}

	// Check mouseover
	public bool CheckMouseOver () {
		if ( mouseRct != new Rect (0,0,0,0) ) {
			return CheckMouseOver ( mouseRct );
		} else {
			return false;
		}	
	}
	
	public bool CheckMouseOver ( Rect rect ) {
		if ( !root ) {
			root = OGRoot.GetInstance();
		}	
		
		Vector2 pos = new Vector2 ( Input.mousePosition.x * root.reverseRatio.x, Input.mousePosition.y * root.reverseRatio.y );

		if ( SystemInfo.deviceType == DeviceType.Handheld && Input.touchCount == 0 ) {
			return false;
		}

		return rect.Contains ( pos );
	}
	
	public bool CheckMouseOver ( Rect rect1, Rect rect2 ) {
		return CheckMouseOver ( rect1 ) || CheckMouseOver ( rect2 );
	}
	
	// Scale (based on screen size)
	public Vector3 RecalcScale () {
		CalcStretch ();
		
		OGScrollView scrollView = this as OGScrollView;

		if ( scrollView ) {
			return new Vector3 ( scrollView.size.x, scrollView.size.y, 1 );
		} else {
			return this.transform.lossyScale;
		}
	}
	
	// Position (based on screen size and flipped vertically because of OpenGL)
	public Vector3 RecalcPosition () {
		CalcAnchor ();
		CalcPivot ();
		
		Vector3 newPos = this.transform.position;
		OGScrollView scrollView = this as OGScrollView;
		
		newPos += offset;
		newPos += scrollOffset;
		
		if ( scrollView ) {
			newPos.y += scrollView.size.y;	
		} else {
			newPos.y += this.transform.lossyScale.y;
		}

		newPos.y = root.screenHeight - newPos.y;
		
		return newPos;
	}
	
	// Stretch (based on screen size)
	private void CalcStretch () {
		if ( stretch == null ) { return; }
		
		Vector3 newScale = this.transform.localScale;
		OGScrollView scrollView = this as OGScrollView;

		if ( scrollView ) {
			newScale.x = scrollView.size.x;
			newScale.y = scrollView.size.y;
		}
		
		if ( stretch.width == ScreenSize.ScreenWidth ) {
			newScale.x = ( root.screenWidth * stretch.widthFactor ) + stretch.widthOffset;
		} else if ( stretch.width == ScreenSize.ScreenHeight ) {
			newScale.x = ( root.screenHeight * stretch.widthFactor ) + stretch.widthOffset;
		}
		
		if ( stretch.height == ScreenSize.ScreenWidth ) {
			newScale.y = ( root.screenWidth * stretch.heightFactor ) + stretch.heightOffset;
		} else if ( stretch.height == ScreenSize.ScreenHeight ) {
			newScale.y = ( root.screenHeight * stretch.heightFactor ) + stretch.heightOffset;
		}
		
		if ( !scrollView ) {
			this.transform.localScale = newScale;
		
		} else {
			scrollView.size.x = newScale.x;
			scrollView.size.y = newScale.y;
		}
	
	}
	
	// Anchor (based on screen size)
	private void CalcAnchor () {
		if ( anchor == null ) { return; }
		
		Vector3 newPos = this.transform.position;
		
		if ( anchor.x != RelativeX.None ) {
			if ( anchor.x == RelativeX.Left ) {
				anchor.xFactor = 0f;
			} else if ( anchor.x == RelativeX.Center ) {
				anchor.xFactor = 0.5f;
			} else if ( anchor.x == RelativeX.Right ) {
				anchor.xFactor = 1f;
			}
			
			newPos.x = (root.screenWidth * anchor.xFactor) + anchor.xOffset;
		}
		
		if ( anchor.y != RelativeY.None ) {
			if ( anchor.y == RelativeY.Top ) {
				anchor.yFactor = 0f;
			} else if ( anchor.y == RelativeY.Center ) {
				anchor.yFactor = 0.5f;
			} else if ( anchor.y == RelativeY.Bottom ) {
				anchor.yFactor = 1f;
			}
			
			newPos.y = (root.screenHeight * anchor.yFactor) + anchor.yOffset;
		}

		this.transform.position = newPos;
	}
	
	// Pivot (based on object size)
	private void CalcPivot () {
		Vector2 scale;
		OGScrollView scrollview = this as OGScrollView; 
		
		if ( scrollview ) {
			scale = scrollview.size;
		} else {
			scale.x = this.transform.lossyScale.x;
			scale.y = this.transform.lossyScale.y;
		}

		switch ( pivot.y ) {
			case RelativeY.Top: case RelativeY.None:
				offset.y = 0;
				break;
								
			case RelativeY.Center:
				offset.y = -scale.y/2;
				break;
				
			case RelativeY.Bottom:
				offset.y = -scale.y;
				break;
		}
		
		switch ( pivot.x ) {	
			case RelativeX.Right:
				offset.x = -scale.x;
				break;
						
			case RelativeX.Center:	
				offset.x = -scale.x/2;
				break;
			
			case RelativeX.Left: case RelativeX.None:
				offset.x = 0;
				break;
		}
	}
	
	// Apply all calculations
	public void Recalculate () {
		if ( !root ) { return; }
		
		OGTexture texture = this as OGTexture;
		Vector3 drawScl = RecalcScale ();
		Vector3 drawPos = RecalcPosition ();
		
		if ( texture != null ) {
			drawCrd.x = 0;
			drawCrd.y = 0;
			drawCrd.width = 1;
			drawCrd.height = 1;
		
		} else if ( currentStyle != null ) {
			drawCrd = currentStyle.coordinates;
		
		} else if ( styles.basic != null ) {
			drawCrd = styles.basic.coordinates;
		}

		if ( isAlwaysOnTop ) {
			drawDepth = 0;
		} else {
			drawDepth = -this.transform.position.z;
		}

		drawRct = new Rect ( drawPos.x, drawPos.y, drawScl.x, drawScl.y );
	}	

	
	//////////////////
	// Init
	//////////////////
	public void Awake () {
		root = OGRoot.GetInstance ();
	}
	
	public void Start () {
		Recalculate ();
	}


	//////////////////
	// Mouse
	//////////////////
	public virtual void OnMouseDown () {}
	public virtual void OnMouseUp () {}
	public virtual void OnMouseOver () {}
	public virtual void OnMouseDrag () {}
	public virtual void OnMouseCancel () {}
	

	//////////////////
	// Update
	//////////////////
	public virtual void UpdateWidget () {} 
	public void ApplyDefaultStyles () {
		if ( !root ) {
			root = OGRoot.GetInstance();
		}
		
		OGSkin skin = root.skin;
		
		if ( !skin ) {
			Debug.LogWarning ( "OpenGUI | No OGSkin attached to OGRoot" );
		} else {
			skin.ApplyDefaultStyles ( this );
		}
	}


	//////////////////
	// Draw
	//////////////////
	public virtual void DrawGL () {}
	public virtual void DrawLine () {}
	public virtual void DrawSkin () {}
	public virtual void DrawText () {}
}
